modules/core/CameraThinLens.cpp

```diff
@@ -0,0 +1,33 @@
+#include "CameraThinLens.h"
+#include "Ray.h"
+#include "random.h"
+#include "Sampler.h"
+#include "macroses.h"
+
+namespace rt
+{
+	void CCameraThinLens::InitRay(Ray& ray, int x, int y, const Vec2f& sample)
+	{
+		CCameraPerspective::InitRay(ray, x, y, sample);
+
+		float lensr = getlensRadius();
+		float nBlades = getnBlades();
+		if (lensr > 0) {
+			// Sample point on lens
+			Vec2f pLens;
+			Vec2f sample(random::U<float>(), random::U<float>());
+			
+			//sample from uniformly distributed points in a regular polygon
+			RT_ASSERT(nBlades == 0 || (nBlades >= 3 && nBlades <= 16));
+			pLens = lensr * CSampler::uniformSampleRegularNgon(sample, nBlades, random::u<int>(1, nBlades));
+			
+			// Compute point on plane of focus
+			float ft = getfocalDistance() / ray.dir.val[2];
```

Comment:
Here I have 2 questions.
1. What if `ray.dir.val[2]` is equal to zero?  (this is a bug)
2. What if `ray.dir.val[2]` is negative? Can then focus distance be negative?

Reply:
I think ray.dir.val[2] will be a zero vector if m_up and m_dir are collinear, so perhaps we should assert that their cross product of m_up and m_dir should be different from zero.

Reply:
2. I think ft shouldn't be negative. Because we just want to use it as a factor to scale the ray so it reaches the focal plane. If it's negative, pFocus (Vec3f pFocus = ray.org + ray.dir * ft;) would be on the opposite side and we don't want that, so I take the absolute value  of ft.

Reply:
I can assure you that the case when `ray.dir.val[2]` is equal or less then zero is very common. 
To check that it is enough to add these lines of code:
`RT_ASSERT(ray.dir.val[2] >= FLT_EPSILON)`
And try different camera directions.

Reply:
> 1. I think ray.dir.val[2] will be a zero vector if m_up and m_dir are collinear, so perhaps we should assert that their cross product of m_up and m_dir should be different from zero.

I think here should be a values calculated from `m_zAxis` to be used. Or, a projection of the ray to the z-Axis.